### DISCUSSION TITLE
Unexpected behaviour of @Valid in RestEasy

### DISCUSSION THREAD

**Discussion by mberdev** [resteasy/resteasy#4211] [Q&A]
I have already asked the question on StackOverflow but I didn't really get an answer: https://stackoverflow.com/questions/78613823/how-to-correctly-apply-valid

TL;DR : `@Valid` needs to be added to **each individual field** of an object in order to get a meaningful message in the HTTP 400. Otherwise the validation still fails (as it should) but we get no message.

It seems to be a different behaviour from, let's say, Spring. And also different from what I read on the web, where it usually says "just put @Valid on a field when its a **nested** object which contents need to be validated too" (which is not the case for a Long).

See paragraph "31.2.2 Validating Entity Data" here : https://docs.oracle.com/javaee/7/tutorial/jaxrs-advanced002.htm
_"To validate these entity classes, use the @Valid annotation **on the method parameter.** "_

**Comment by jamezp** [comment]
What version of RESTEasy are you using?

**Reply by mberdev** [reply]
**RestEasy 4.7.6**

The problem was not directly caused by RestEasy (or was it?), **but in any case it's awkward.**

The root cause was that RestEasy's validation was running, but the _actual result_ didn't matter because the final instantiation of the object was failing. It was failing on a technicality : The class to instantiate was not marked "static" inside the test class and it seemed to bother RestEasy's automated population of constructor parameters. RestEasy wants a "proper" class. I discovered that by accident in the logs, after spending a lot of time trying many validation scenarios.

In other words : you get bad request regardless of the "real" validation just because `new MyPayloadClass(...)` failed at the very end -- and not even for a very good reason (at least from where I stand).

Do what you want with this information.

**Reply by jamezp** [reply]
Hmm.. ...so does the error seem to happen when RESTEasy attempts to instantiate something?

**Reply by mberdev** [reply]
> Hmm.. ...so does the error seem to happen when RESTEasy attempts to instantiate something?

Yes, that's what I wrote. The instantiation just after sucessfully deserializing and validating the data.

I think it's by design, **but it is nonetheless awkward.**


Reproducible example : 

This returns **HTTP 200 (OK) , as it should**

```
    @Value
    public static class PayloadStatic{
        Long mSomeLong;
    }

    @POST
    @Path("my-test")
    @Produces(MediaType.APPLICATION_JSON)
    public Response myTest(
        final @Valid PayloadStaticpayload
    ) {
        return ok();
    }

```
body json :

```
{
    "someLong": null
}

```
This confusingly returns **HTTP 400 (bad request)**

```

    @Value
    public class PayloadNotStatic{
        Long mSomeLong;
    }

    @POST
    @Path("my-test")
    @Produces(MediaType.APPLICATION_JSON)
    public Response myTest(
        final @Valid PayloadStaticpayload
    ) {
        return ok();
    }

```

Same json. The only difference is keyword "static"


Server error : 

**Cannot construct instance of `admin.endpoint.AnnotationsEndpoint$PayloadNotStatic`: non-static inner classes like this can only by instantiated using default, no-argument constructor**

```

2024-06-19 09:51:57,234 WARNING [common.UncaughtExceptionHandler] (default task-771) Invalid JSON format: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `admin.endpoint.AnnotationsEndpoint$PayloadNotStatic`: non-static inner classes like this can only by instantiated using default, no-argument constructor
 at [Source: (io.undertow.servlet.spec.ServletInputStreamImpl); line: 2, column: 5]
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1764)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:400)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1204)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1412)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:363)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:196)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:322)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:2007)
	at com.fasterxml.jackson.core.jackson-databind@2.12.7.1//com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1174)
...
```

_Technically_ the error comes from fasterXML, but I guess it also comes from the way RestEasy uses it.

**Reply by jamezp** [reply]
I'm not too sure what RESTEasy could do about that since it delegates to Jackson. *Maybe* it would with Jakarta JSON Binding instead, I'm not too sure.

**Reply by mberdev** [reply]
> I'm not too sure what RESTEasy could do about that since it delegates to Jackson. _Maybe_ it would with Jakarta JSON Binding instead, I'm not too sure.

It could catch this specific exception and transform it into an HTTP 500. Because despite what the message says, it has nothing to do with Json format and everything to do with the Java side of things.

**Reply by jamezp** [reply]
I really hope that I can find some time this week to create a reproducer so I can better understand what is happening here.
